File: problems/Q2/Q2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
const {createLinkedList, ListNode, printLinkedList} = require('../helpers/linkedlist/ListNode.js');

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let res = new ListNode()
    let digit = null
    let carry = 0
    while (l1 || l2 || carry) {
        if (!digit) digit = res
        else {
            digit.next = new ListNode()
            digit = digit.next
        }
        let value = carry + (l1?l1.val:0) + (l2?l2.val:0)
        l1 = l1?l1.next:0
        l2 = l2?l2.next:0
        carry = Math.floor(value / 10)
        digit.val = value % 10
    }
    return res
};

printLinkedList(addTwoNumbers(
                    createLinkedList([2, 4, 3]),
                    createLinkedList([5, 6, 4])
                ))

File: problems/Q2/Q2.lua
package.path = '../helpers/linkedlist/ListNode.lua;'..package.path
require 'ListNode'

local function addTwoNumbers(l1, l2)
    local res = ListNode({})
    local digit = nil
    local carry = 0
    while l1 or l2 or carry ~= 0 do
        if not digit then
            digit = res
        else
            digit.next_node = ListNode({})
            digit = digit.next_node
        end
        value = carry + (l1 and l1.val or 0) + (l2 and l2.val or 0)
        l1 = l1 and l1.next_node
        l2 = l2 and l2.next_node
        carry = math.floor( value / 10 )
        digit.val = value % 10
    end
    return res
end

printLinkedList(
    addTwoNumbers(
        createLinkedList({2,4,3}),
        createLinkedList({5,6,4})
    )
)